src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java

```diff
@@ -0,0 +1,200 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom;
+
+import java.io.File;
+
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.dsl.RepositoryHandler;
+import org.gradle.api.artifacts.repositories.IvyArtifactRepository;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.invocation.Gradle;
+import org.gradle.api.plugins.PluginAware;
+
+public class LoomRepositoryPlugin implements Plugin<PluginAware> {
+	@Override
+	public void apply(PluginAware target) {
+		RepositoryHandler repositories = null;
+
+		if (target instanceof Settings settings) {
+			repositories = settings.getDependencyResolutionManagement().getRepositories();
+
+			// leave a marker so projects don't try to override these
+			settings.getGradle().getPluginManager().apply(LoomRepositoryPlugin.class);
+		} else if (target instanceof Project project) {
+			if (project.getGradle().getPlugins().hasPlugin(LoomRepositoryPlugin.class)) {
+				return;
+			}
+
+			repositories = project.getRepositories();
+		} else if (target instanceof Gradle) {
+			return;
+		} else {
+			throw new IllegalArgumentException("Expected target to be a Project or Settings, but was a " + target.getClass());
+		}
+
+		Cache cache = new Cache(target);
+
+		// MavenConfiguration.java
+		repositories.flatDir(repo -> {
+			repo.setName("UserLocalCacheFiles");
+			repo.dir(cache.getRootBuildCache());
+		});
+		repositories.maven(repo -> {
+			repo.setName("UserLocalRemappedMods");
+			repo.setUrl(cache.getRemappedModCache());
+		});
+		repositories.maven(repo -> {
+			repo.setName("Fabric");
+			repo.setUrl("https://maven.fabricmc.net/");
+		});
+		repositories.maven(repo -> {
+			repo.setName("Mojang");
+			repo.setUrl("https://libraries.minecraft.net/");
+		});
+		repositories.mavenCentral();
+
+		// MinecraftMappedProvider.java
+		repositories.ivy(repo -> {
+			repo.setUrl(cache.getUserCache());
+			repo.patternLayout(layout -> {
+				layout.artifact("[revision]/[artifact]-[revision](.[ext])");
+			});
+			repo.metadataSources(IvyArtifactRepository.MetadataSources::artifact);
+		});
+
+		// MinecraftProcessedProvider.java
+		repositories.ivy(repo -> {
+			repo.setUrl(cache.getRootPersistentCache());
+			repo.patternLayout(layout -> {
+				layout.artifact("[revision]/[artifact]-[revision](.[ext])");
+			});
+			repo.metadataSources(IvyArtifactRepository.MetadataSources::artifact);
+		});
+	}
+}
+
+final class Cache {
```

Comment:
Might be out of scope for this PR, but it does duplicate some stuff in LoomGradleExtension, moving this out into its own class and removing the duplicates might not be a bad idea. 

At somepoint I want to give LGE a good cleanup so im not too fussed about doing this now.